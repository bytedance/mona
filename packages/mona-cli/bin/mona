#! /usr/bin/env node

const mona = require('../dist').default;
mona();
