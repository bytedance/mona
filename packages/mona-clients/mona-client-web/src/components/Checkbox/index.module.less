.checkbox {
  -webkit-tap-highlight-color: transparent;
  display: inline-block;

  & .wrapper {
    display: -webkit-inline-flex;
    display: inline-flex;
    -webkit-align-items: center;
    align-items: center;
    vertical-align: middle;
  }

  & .input {
    margin-right: 5px;
    -webkit-appearance: none;
    appearance: none;
    outline: 0;
    background-color: #FFF;
    border: 1px solid #E8E8E8;
    border-radius: 100%;
    width: 22px;
    height: 22px;
    position: relative;
  }

  & .input.checked:before {
    font: normal normal normal 14px/1 "ttui";
    content: "\e601";
    font-size: 28px;
    color: #FFFFFF;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -48%) scale(0.73);
    -webkit-transform: translate(-50%, -48%) scale(0.73);
  }
}