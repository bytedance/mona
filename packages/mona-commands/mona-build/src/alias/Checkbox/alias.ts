import { bubbleEventsAlias, basePropAlias, propAliasMap } from '../prop';
import { CheckboxAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  disabled: propAliasMap.disabled,
  checked: propAliasMap.checked,
  color: propAliasMap.color,
  value: propAliasMap.value,
};

export default alias;
