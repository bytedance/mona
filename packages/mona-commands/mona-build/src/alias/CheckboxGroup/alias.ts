import { bubbleEventsAlias, basePropAlias, propAliasMap } from '../prop';
import { CheckboxGroupAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  onChange: propAliasMap.disabled,
  name: propAliasMap.name,
};

export default alias;
