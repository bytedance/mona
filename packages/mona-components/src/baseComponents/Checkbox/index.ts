import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps } from '../base';

export interface CheckboxProps extends BaseProps {
  value?: string
  disabled?: boolean;
  checked?: boolean
  color?: string
}

export const Checkbox: React.ComponentType<CheckboxProps> = createComponent<CheckboxProps>('checkbox')

