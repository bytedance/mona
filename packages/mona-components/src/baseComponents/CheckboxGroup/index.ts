import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps, EventHandler } from '../base';

export interface CheckboxGroupProps extends BaseProps {
  onChange?: EventHandler;
  name?: string;
}

export const CheckboxGroup: React.ComponentType<CheckboxGroupProps> = createComponent<CheckboxGroupProps>('checkbox-group')

