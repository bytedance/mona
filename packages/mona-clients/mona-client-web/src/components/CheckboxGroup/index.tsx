import React, { useRef } from 'react';
import { CheckboxGroupProps, TouchEvent } from '@bytedance/mona';
import { useHandlers } from '../hooks';

export const EMPTY_ITEM = Symbol('checkboxEmpty');

interface CheckboxGroupContextProps {
  initValue: (value?: string, checked?: boolean) => number;
  changeValue: (index: number, value?: string) => void
  toggleChecked: (index: number, checked?: boolean, e?: TouchEvent) => void
}

interface ValueItem {
  value?: string;
  checked?: boolean;
}

export const CheckboxGroupContext = React.createContext<CheckboxGroupContextProps | null>(null)

const CheckboxGroup: React.FC<CheckboxGroupProps> = ({ children, onChange, ...restProps }) => {
  const { handleClassName, ...handlerProps } = useHandlers(restProps);
  // const [values, setValues] = useState<ValueItem[]>([]);
  const valuesRef = useRef<ValueItem[]>([])
  const indexRef = useRef(0);
  
  const context = {
    initValue: (value?: string, checked?: boolean) => {
      const index = indexRef.current++;
      const newValues = [...valuesRef.current];
      // record order
      newValues[index] = { value, checked };
      valuesRef.current = newValues;
      return index;
    },
    changeValue: (index: number, value?: string) => {
      const newValues = [...valuesRef.current];
      newValues[index] = { ...newValues[index], value };
      valuesRef.current = newValues;
    },
    toggleChecked: (index: number, checked?: boolean, event?: TouchEvent) => {
      const newValues = [...valuesRef.current];
      newValues[index] = { ...newValues[index], checked }
      valuesRef.current = newValues;
      
      if (typeof onChange === 'function') {
        const result = newValues.filter(v => v.checked).map(v => v.value);
        const e = { ...event!, detail: { value: result as string[] } };
        onChange(e)
      }
    }
  }

  return (
    <div className={handleClassName()} {...handlerProps}>
      <CheckboxGroupContext.Provider value={context}>
        {children}
      </CheckboxGroupContext.Provider>
    </div>
  )
}

export default CheckboxGroup;
