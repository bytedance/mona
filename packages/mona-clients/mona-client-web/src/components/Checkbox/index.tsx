import React, { useState, useEffect, useContext, useRef } from 'react';
import { CheckboxProps } from '@bytedance/mona';
import cs from 'classnames';
import { CheckboxGroupContext } from '../CheckboxGroup';
import styles from './index.module.less';
import { useHandlers } from '../hooks';
import { formatMouseEvent } from '../utils';

const Checkbox: React.FC<CheckboxProps> = (props) => {
  const {
    value,
    disabled,
    checked,
    color = '#F85959',
    children,
    id,
    ...restProps
  } = props;

  const group = useContext(CheckboxGroupContext);
  const orderRef = useRef(-1);

  const { handleClassName, onClick, ...handlerProps } = useHandlers(restProps);

  const [isChecked, setIsChecked] = useState(checked);

  const handleClick = (e: React.MouseEvent) => {
    const newIsChecked = !isChecked;
    setIsChecked(newIsChecked);
    if (orderRef.current >= 0) {
      group?.toggleChecked(orderRef.current, newIsChecked, formatMouseEvent({ event: e, type: 'change' }));
    }
    onClick(e);
  }
  
  // don't change order of useEffects, the order is important
  useEffect(() => {
    setIsChecked(checked);
    if (orderRef.current >= 0) {
      group?.toggleChecked(orderRef.current, checked);
    }
  }, [checked])

  useEffect(() => {
    orderRef.current = group ? group?.initValue(value, checked) : -1;
  }, [])

  useEffect(() => {
    if (orderRef.current >= 0) {
      group?.changeValue(orderRef.current, value);
    }
  }, [value])

  return (
    <div
      className={handleClassName(styles.checkbox)}
      {...handlerProps}
    >
      <div className={styles.wrapper}>
        <div className={cs(styles.input, {[styles.checked]: isChecked})} style={{ borderColor: color, backgroundColor: isChecked ? color: '' }}>
          <input value={value} id={id} onClick={handleClick} type="checkbox" className={styles.inner}></input>
        </div>
        {children}
      </div>
    </div>
  )
}

export default Checkbox;
