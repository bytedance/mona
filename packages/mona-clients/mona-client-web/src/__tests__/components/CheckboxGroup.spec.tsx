import React from 'react';
import { mount } from 'enzyme';
import CheckboxGroup from '../../components/CheckboxGroup';
import mountTest from '../../../../../tests/shared/mountTest';
import handlerTest from '../../../../../tests/shared/handlerTest';
import { CheckboxGroupProps } from '@bytedance/mona';
import Checkbox from '../../components/Checkbox';

describe('web component: CheckboxGroup', () => {
  mountTest(CheckboxGroup)
  handlerTest({ Component: CheckboxGroup });

  it('should render correctly', () => {
    expect(mount((
      <CheckboxGroup>
        <Checkbox value="apple">apple</Checkbox>
        <Checkbox value="banana">banana</Checkbox>
        <Checkbox value="orange">orange</Checkbox>
      </CheckboxGroup>
    )).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: CheckboxGroupProps = {
      className: 'class-name',
      name: 'group',
      style: { color: 'red' },
      onChange: () => {}
    }
    const wrapper = mount(<Checkbox {...props}>hello</Checkbox>);
    expect(wrapper.render()).toMatchSnapshot();
  })

  it('should trigger onChange when tap any option', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <CheckboxGroup onChange={onChange}>
        <Checkbox value="apple">apple</Checkbox>
        <Checkbox value="banana">banana</Checkbox>
        <Checkbox value="orange">orange</Checkbox>
      </CheckboxGroup>
    )

    wrapper.find('input[value="banana"]').first().simulate('click');
    expect(onChange.mock.calls[0][0].detail.value).toEqual(['banana']);
    wrapper.find('input[value="apple"]').first().simulate('click');
    expect(onChange.mock.calls[1][0].detail.value).toEqual(['apple', 'banana']);
    wrapper.find('input[value="apple"]').first().simulate('click');
    expect(onChange.mock.calls[2][0].detail.value).toEqual(['banana']);
  }) 
})
