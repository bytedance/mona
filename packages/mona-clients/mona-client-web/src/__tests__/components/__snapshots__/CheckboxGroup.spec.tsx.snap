// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: CheckboxGroup should have right properties 1`] = `
LoadedCheerio {
  "0": <div
    class="checkbox class-name  "
    name="group"
    style="color: red;"
  >
    <div
      class="wrapper"
    >
      <div
        class="input "
        style="border-color: #f85959;"
      >
        <input
          class="inner"
          type="checkbox"
          value=""
        />
      </div>
      hello
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: CheckboxGroup should render correctly 1`] = `
LoadedCheerio {
  "0": <div
    class="   "
  >
    <div
      class="checkbox   "
    >
      <div
        class="wrapper"
      >
        <div
          class="input "
          style="border-color: #f85959;"
        >
          <input
            class="inner"
            type="checkbox"
            value="apple"
          />
        </div>
        apple
      </div>
    </div>
    <div
      class="checkbox   "
    >
      <div
        class="wrapper"
      >
        <div
          class="input "
          style="border-color: #f85959;"
        >
          <input
            class="inner"
            type="checkbox"
            value="banana"
          />
        </div>
        banana
      </div>
    </div>
    <div
      class="checkbox   "
    >
      <div
        class="wrapper"
      >
        <div
          class="input "
          style="border-color: #f85959;"
        >
          <input
            class="inner"
            type="checkbox"
            value="orange"
          />
        </div>
        orange
      </div>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
